subprojects/core-api/src/main/java/org/gradle/api/execution/incremental/InputChanges.java

```diff
@@ -0,0 +1,98 @@
+import org.gradle.api.tasks.incremental.InputFileDetails;
+
+/**
+ * Provides access to any input files that need to be processed by an incremental work action.
+ *
+ * <p>
+ * An incremental work action is one that accepts a single {@link InputChanges} parameter.
+ * The work action can then query what changed for an input property since the last execution to only process the changes.
+ *
+ * <pre class='autoTested'>
+ * class IncrementalReverseTask extends DefaultTask {
+ *     {@literal @}InputDirectory
+ *     def File inputDir
+ *
+ *     {@literal @}OutputDirectory
+ *     def File outputDir
+ *
+ *     {@literal @}TaskAction
+ *     void execute(InputChanges inputChanges) {
+ *         if (!inputChanges.incremental)
+ *             project.delete(outputDir.listFiles())
+ *
+ *         inputChanges.getFileChanges(inputDir).each { change -&gt;
+ *             if (change.removed) {
+ *                 def targetFile = project.file("$outputDir/${change.file.name}")
+ *                 if (targetFile.exists()) {
+ *                     targetFile.delete()
+ *                 }
+ *             } else {
+ *                 def targetFile = project.file("$outputDir/${change.file.name}")
+ *                 targetFile.text = change.file.text.reverse()
+ *             }
+ *         }
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>
+ * In the case where Gradle is unable to determine which input files need to be reprocessed, then all of the input files will be reported as added.
+ * Cases where this occurs include:
+ * <ul>
+ *     <li>There is no history available from a previous execution.</li>
+ *     <li>An non-file input property has changed since the previous execution.</li>
+ *     <li>One or more output files have changed since the previous execution.</li>
+ * </ul>
+ *
+ * @since 5.4
+ */
+@NonExtensible
+@Incubating
+public interface InputChanges {
+    /**
+     * Indicates if it was possible for Gradle to determine which input files were out of date compared to a previous execution.
+     * Incremental inputs are unavailable when history is unavailable (i.e. this piece of work has never been executed before), or if there are changes to non-file input properties, or output files.
+     * <p>
+     * When <code>true</code>:
+     * </p>
+     * <ul>
+     *     <li>{@link #getFileChanges(Object)} reports changes to the input files compared to the previous execution.</li>
+     * </ul>
+     * <p>
+     * When <code>false</code>:
+     * </p>
+     * <ul>
+     *     <li>Every input file is reported via {@link #getFileChanges(Object)} as if it was 'added'.</li>
+     * </ul>
+     */
+    boolean isIncremental();
+
+    /**
+     * Changes for the property.
+     *
+     * <p>When {@link #isIncremental()} is {@code false}, then all elements of the property are returned as added.</p>
+     *
+     * @param property The instance of the property to query.
+     */
+    Iterable<InputFileDetails> getFileChanges(Object property);
```

Comment:
Should we consider using `Stream<>` here?

Reply:
We could. When using stream, we could add some logic later to only produce the changes which are consumed if we wanted to.

What other benefits do we get from returning `Stream` here?

Reply:
We could do lazy like that with `Iterable<>`, too. I'm not sure if `Stream<>` has any additional benefits. Maybe it's harder to use from Kotlin? Or simpler?

Reply:
One of the benefits of `Stream` is that you can only iterate it once. That makes it clearer for the user of the API that they should only do it once.